sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java

```diff
@@ -52,280 +59,496 @@
+
+      private static PipelineOptions dummyRunner() {
+        final PipelineOptions pipelineOptions = PipelineOptionsFactory.create();
+        pipelineOptions.setRunner(DummyRunner.class);
+        return pipelineOptions;
+      }
+
+      @Category(RunnableOnService.class)
+      @Test
+      public void testNormalFlow() throws Exception {
+        addTransform(pCollection(pipeline));
+        pipeline.run();
+      }
+
+      @Category(RunnableOnService.class)
+      @Test
+      public void testMissingRun() throws Exception {
+
+        exception.expect(TestPipeline.PipelineRunMissingException.class);
+
+        addTransform(pCollection(pipeline));
+      }
+
+      @Category(RunnableOnService.class)
+      @Test
+      public void testMissingRunWithDisabledEnforcement() throws Exception {
+        pipeline.enableAbandonedNodeEnforcement(false);
+        addTransform(pCollection(pipeline));
+
+        // disable abandoned node detection
+      }
+
+      @Category(RunnableOnService.class)
+      @Test
+      public void testMissingRunAutoAdd() throws Exception {
+        pipeline.enableAutoRunIfMissing(true);
+        addTransform(pCollection(pipeline));
+
+        // have the pipeline.run() auto-added
+      }
+
+      @Category(RunnableOnService.class)
+      @Test
+      public void testDanglingPTransformRunnableOnService() throws Exception {
+
+        exception.expect(TestPipeline.AbandonedNodeException.class);
+        exception.expectMessage(P_TRANSFORM);
+
+        final PCollection<String> pCollection = pCollection(pipeline);
+        PAssert.that(pCollection).containsInAnyOrder(WHATEVER);
+        pipeline.run().waitUntilFinish();
+
+        // dangling PTransform
+        addTransform(pCollection);
+      }
+
+      @Category(NeedsRunner.class)
+      @Test
+      public void testDanglingPTransformNeedsRunner() throws Exception {
+
+        exception.expect(TestPipeline.AbandonedNodeException.class);
+        exception.expectMessage(P_TRANSFORM);
+
+        final PCollection<String> pCollection = pCollection(pipeline);
+        PAssert.that(pCollection).containsInAnyOrder(WHATEVER);
+        pipeline.run().waitUntilFinish();
+
+        // dangling PTransform
+        addTransform(pCollection);
+      }
+
+      @Category(RunnableOnService.class)
+      @Test
+      public void testDanglingPAssertRunnableOnService() throws Exception {
+
+        exception.expect(TestPipeline.AbandonedNodeException.class);
+        exception.expectMessage(P_ASSERT);
+
+        final PCollection<String> pCollection = pCollection(pipeline);
+        PAssert.that(pCollection).containsInAnyOrder(WHATEVER);
+        pipeline.run().waitUntilFinish();
+
+        // dangling PAssert
+        PAssert.that(pCollection).containsInAnyOrder(WHATEVER);
+      }
+
+      @Category(RunnableOnService.class)
+      @Test
+      public void testNoTestPipelineUsedRunnableOnService() { }
```

Comment:
This is a really strange test - It's at least worth commenting what it's testing more explicitly (e.g. 
"The `{@code TestPiplieline}` is initialized as a field in the test class, but is not used in this test; with the default enforcement active, this test succeeds".

The same is true of the test immediately following, and any other empty test.